### PR TITLE
Clear methods when Gunicorn worker exits

### LINKED ISSUE #2873: Make custom metrics work with gunicorn reload
Right now the custom metrics are exposed through a process in the model container, since the metrics contains label `worker-id`, each restart/reload of the gunicorn will create a new data series, this caused an issue, as both old and new data series are exists and will be send back as prometheus scrape response.

With auto-reload process in our model (in order to avoid potential OOM), this resulting over 10MB prometheus response after running the model for a while, which caused scrape timeout and huge memory bump of the prometheus server.

I think we need a better way to workaround this issue, as reload gunicorn seems a common method to avoid OOM in model deployment.

### PR DESCRIPTION
**Pull request by adriangonz**
**What this PR does / why we need it**:

Clear the internal Python server metrics when a Gunicorn worker exists. Otherwise, the internal dictionary keeps growing, causing a high memory usage in the server itself and other downstream issues in Prometheus. 

**Which issue(s) this PR fixes**:

Fixes #2873

**Special notes for your reviewer**:

Only the entries relevant to a worker process will be cleared. 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **adriangonz**
You can assign the PR to them by writing `/assign @adriangonz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: python/seldon_core/metrics.py
Comment (RafalSkolasinski) [inline python/seldon_core/metrics.py:192]: hmm... is the `#{...}` some new thing? shouldn't it be `f-string`?

File: python/tests/test_gunicorn_utils.py
Comment (RafalSkolasinski) [inline python/tests/test_gunicorn_utils.py:12]: Where does port `6005` comes from?

File: python/tests/test_gunicorn_utils.py
Comment (RafalSkolasinski) [inline python/tests/test_gunicorn_utils.py:42]: Should we maybe have a test that kills only one worker as well?

**Review by RafalSkolasinski — APPROVED** [review]
Nice one! 

Just have a few minor questions, see inlines.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/3018#pullrequestreview-605044936 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by RafalSkolasinski** [comment]
Also the lint is failing

File: python/seldon_core/metrics.py
Comment (adriangonz) [reply python/seldon_core/metrics.py:192]: Good catch. Let me update that.

File: python/tests/test_gunicorn_utils.py
Comment (adriangonz) [reply python/tests/test_gunicorn_utils.py:12]: That's defined in `python/tests/helpers.py`

File: python/tests/test_gunicorn_utils.py
Comment (adriangonz) [reply python/tests/test_gunicorn_utils.py:42]: That's slightly more complex. I think it falls out of scope. This is the usual way to restart workers in Gunicorn as far as I can see.

**Comment by anggao** [comment]
Thanks! This looks great I will test it next week!

**Comment by seldondev** [comment]
@adriangonz: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | af97edadaa6bb060fe8e954d9a75fad618449378 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-hfkrr) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@adriangonz: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | af97edadaa6bb060fe8e954d9a75fad618449378 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-hfkrr) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by RafalSkolasinski** [comment]
For some reason tests were again triggered twice. I see one set of integration and notebooks to passed:
image

**Comment by seldondev** [comment]
@adriangonz: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | af97edadaa6bb060fe8e954d9a75fad618449378 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-hfkrr) | `/test notebooks`
integration | af97edadaa6bb060fe8e954d9a75fad618449378 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/integration-pwsmv) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@adriangonz: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | af97edadaa6bb060fe8e954d9a75fad618449378 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-hfkrr) | `/test notebooks`
integration | af97edadaa6bb060fe8e954d9a75fad618449378 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/integration-pwsmv) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Pull request merged**
